File: src/app/plugin/binding-chip/chip.cpp
/**
 *
 *    Copyright (c) 2020 Project CHIP Authors
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

/**
 *    @file
 *      This file provides the CHIP implementation of functions required
 *      by the CHIP ZCL Application Layer
 *
 */

#include "chip-zcl.h"

#include <support/logging/CHIPLogging.h>
#include <system/SystemPacketBuffer.h>

#include <stdarg.h>

void chipZclCorePrintln(const char * formatString, ...)
{
    va_list args;
    va_start(args, formatString);
    chip::Logging::LogV(chip::Logging::kLogModule_Zcl, chip::Logging::kLogCategory_Detail, formatString, args);
    va_end(args);
}

// Default Response Stubs
ChipZclStatus_t chipZclSendDefaultResponse(const ChipZclCommandContext_t * context, ChipZclStatus_t status)
{
    return CHIP_ZCL_STATUS_SUCCESS;
}

// Reporting Configuration Stubs
void chipZclReportingConfigurationsFactoryReset(ChipZclEndpointId_t endpointId)
{
    return;
}

// Endpoint Management Stubs
ChipZclEndpointId_t chipZclEndpointIndexToId(ChipZclEndpointIndex_t index, const ChipZclClusterSpec_t * clusterSpec)
{
    return CHIP_ZCL_ENDPOINT_NULL;
}

// Attribute Management Stubs
void chipZclResetAttributes(ChipZclEndpointId_t endpointId)
{
    return;
}

// Event Management Stubs
EventQueue emAppEventQueue;

Event * chipZclEventFind(EventQueue * queue, EventActions * actions, EventPredicate predicate, void * data, bool all)
{
    return 0;
}

void chEventControlSetDelayMS(ChipZclEventControl * event, uint32_t delay)
{
    return;
}

void chipZclEventSetDelayMs(Event * event, uint32_t delay)
{
    return;
}

int32_t chipZclCompareClusterSpec(const ChipZclClusterSpec_t * s1, const ChipZclClusterSpec_t * s2)
{
    return 0;
}

bool chipZclAreClusterSpecsEqual(const ChipZclClusterSpec_t * s1, const ChipZclClusterSpec_t * s2)
{
    return 0;
}

void chipZclReverseClusterSpec(const ChipZclClusterSpec_t * s1, ChipZclClusterSpec_t * s2)
{
    return;
}

/**
 * ChipZclBuffer_t is merely a type alias
 */
struct ChipZclBuffer_t : public chip::System::PacketBuffer
{
};

/**
 * Function that allocates a buffer.
 */
ChipZclBuffer_t * chipZclBufferAlloc(uint16_t allocatedLength)
{
    return static_cast<ChipZclBuffer_t *>(chip::System::PacketBuffer::NewWithAvailableSize(allocatedLength));
}

/**
 * Function that frees a buffer and its storage.
 */
void chipZclBufferFree(ChipZclBuffer_t * buffer)
{
    chip::System::PacketBuffer::Free(buffer);
}

/**
 * Function that returns a pointer to the raw buffer.
 */
uint8_t * chipZclBufferPointer(ChipZclBuffer_t * buffer)
{
    return buffer->Start();
}

/**
 * Function that returns the size of the used portion of the buffer.
 */
uint16_t chipZclBufferDataLength(ChipZclBuffer_t * buffer)
{
    return buffer->DataLength();
}

/**
 * returns space available for writing after any data already in the buffer
 */
uint16_t chipZclBufferAvailableLength(ChipZclBuffer_t * buffer)
{
    return buffer->AvailableDataLength();
}

/**
 *  sets data length for a buffer that's being written to
 */
void chipZclBufferSetDataLength(ChipZclBuffer_t * buffer, uint16_t newLength)
{
    buffer->SetDataLength(newLength);
}

File: src/test_driver/happy/lib/Chip.py
#!/usr/bin/env python3

#
#    Copyright (c) 2020 Project CHIP Authors
#    Copyright (c) 2015-2017 Nest Labs, Inc.
#    All rights reserved.
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
#

##
#    @file
#       Implements CHIP class that wraps around standalone CHIP code library.
#

import os
import sys

from happy.Utils import *
from ChipState import ChipState


class Chip(ChipState):
    def __init__(self):
        ChipState.__init__(self)

        self.chip_happy_conf_path = None
        self.chip_build_path = None

        self.max_running_time = 1800

    def __check_chip_path(self):
        # Pick chip path from configuration
        if "chip_path" in self.configuration.keys():
            self.chip_happy_conf_path = self.configuration["chip_path"]
            emsg = "Found chip path: %s." % (self.chip_happy_conf_path)
            self.logger.debug("[localhost] Chip: %s" % (emsg))

        # Check if Chip build path is set
        if "TEST_BIN_DIR" in os.environ.keys():
            self.chip_build_path = os.environ['TEST_BIN_DIR']
            emsg = "Found chip TEST_BIN_DIR: %s." % (self.chip_build_path)
            self.logger.debug("[localhost] Chip: %s" % (emsg))

        if self.chip_build_path is not None:
            self.chip_path = self.chip_build_path
        else:
            self.chip_path = self.chip_happy_conf_path

        if self.chip_path is None:
            emsg = "Unknown path to Chip directory (repository)."
            self.logger.error("[localhost] Chip: %s" % (emsg))
            self.logger.info(
                "Set chip_path with happy-configuration and try again.")
            sys.exit(1)

        if not os.path.exists(self.chip_path):
            emsg = "Chip path %s does not exist." % (self.chip_path)
            self.logger.error("[localhost] Chip: %s" % (emsg))
            self.logger.info(
                "Set correct chip_path with happy-configuration and try again.")
            sys.exit(1)

        if self.chip_path[-1] == "/":
            self.chip_path = self.chip_path[:-1]

    def __get_cmd_path(self, cmd_end):
        cmd_path = self.chip_path + "/" + str(cmd_end)
        if not os.path.exists(cmd_path):
            emsg = "Chip path %s does not exist." % (cmd_path)
            self.logger.error("[localhost] Chip: %s" % (emsg))
            sys.exit(1)
            # return None
        else:
            return cmd_path

    def getChipInetLayerMulticastPath(self):
        self.__check_chip_path()
        cmd_path = self.__get_cmd_path("TestInetLayerMulticast")
        return cmd_path
